SentinelOne/sentinelone_module/singularity/connectors.py

```diff
@@ -0,0 +1,158 @@
+import asyncio
+import time
+from abc import ABC
+from datetime import datetime, timedelta, timezone
+from functools import cached_property
+from typing import Any, Optional
+
+import orjson
+from dateutil.parser import isoparse
+from loguru import logger
+from sekoia_automation.aio.connector import AsyncConnector
+from sekoia_automation.connector import Connector, DefaultConnectorConfiguration
+from sekoia_automation.storage import PersistentJSON
+
+from sentinelone_module.base import SentinelOneModule
+from sentinelone_module.metrics import EVENTS_LAG, FORWARD_EVENTS_DURATION, OUTCOMING_EVENTS
+from sentinelone_module.singularity.client import SingularityClient
+
+
+class SingularityConnectorConfig(DefaultConnectorConfiguration):
+    """Configuration for SingularityConnector."""
+
+    frequency: int = 60
+
+
+class AbstractSingularityConnector(AsyncConnector, ABC):
+    module: SentinelOneModule
+    configuration: SingularityConnectorConfig
+
+    product_name: str
+
+    def __init__(self, *args: Any, **kwargs: Optional[Any]) -> None:
+        super().__init__(*args, **kwargs)
+        self.context = PersistentJSON(f"{self.product_name}_context.json", self._data_path)
+
+    @cached_property
+    def client(self) -> SingularityClient:
+        return SingularityClient(
+            hostname=self.module.configuration.hostname,
+            api_token=self.module.configuration.api_token,
+        )
+
+    def stop(self, *args: Any, **kwargs: Optional[Any]) -> None:
+        """
+        Stop the connector.
+
+        Temporary redefine the method to avoid known SDK issues.
+        """
+        super(Connector, self).stop(*args, **kwargs)
+
+    @property
+    def last_event_date(self) -> datetime:
+        """
+        Get last event date.
+
+        Returns:
+            datetime:
+        """
+        now = datetime.now(timezone.utc)
+        one_week_ago = (now - timedelta(days=7)).replace(microsecond=0)
+
+        with self.context as cache:
+            last_event_date_str = cache.get("last_event_date")
+
+            # If undefined, retrieve events from the last 7 days
+            if last_event_date_str is None:
+                return one_week_ago
+
+            # Parse the most recent date seen
+            last_event_date = isoparse(last_event_date_str)
+
+            # We don't retrieve messages older than one week
+            if last_event_date < one_week_ago:
+                return one_week_ago
+
+            return last_event_date
+
+    @property
+    def last_checkpoint(self) -> str | None:
+        with self.context as cache:
+            result: str | None = cache.get("last_checkpoint")
+
+            return result
+
+    async def single_run(self) -> int:
+        result = 0
+
+        while True:
```

Comment:
I suggest using `self.running` here.
```suggestion
        while self.running:
```